domains/primitives/runtime/src/lib.rs

```diff
@@ -195,20 +216,52 @@ sp_api::decl_runtime_apis! {
         /// Returns true if the extrinsic is an inherent extrinsic.
         fn is_inherent_extrinsic(extrinsic: &<Block as BlockT>::Extrinsic) -> bool;
 
-        /// Checks the validity of extrinsic in a bundle.
+        /// Checks the validity of extrinsic.
+        #[changed_in(2)]
+        fn check_transaction_validity(
+            uxt: &<Block as BlockT>::Extrinsic,
+            block_number: NumberFor<Block>,
+            block_hash: <Block as BlockT>::Hash
+        ) -> Result<(), CheckTxValidityError>;
+
+        /// Checks the validity of extrinsic
+        /// v2 can check validity of unsigned and self contained extrinsics as well.
         fn check_transaction_validity(
             uxt: &<Block as BlockT>::Extrinsic,
             block_number: NumberFor<Block>,
             block_hash: <Block as BlockT>::Hash
         ) -> Result<(), CheckTxValidityError>;
 
+        /// New in v2.
+        /// Checks transaction and also perform pre dispatch on transaction.
+        /// This is useful to maintain side effect in the RuntimeApi instance's storage buffer.
+        fn check_transaction_and_do_pre_dispatch(
+            uxt: &<Block as BlockT>::Extrinsic,
+            block_number: NumberFor<Block>,
+            block_hash: <Block as BlockT>::Hash
+        ) -> Result<(), CheckTxValidityError>;
+
         /// Returns the storage keys of states accessed in the API `check_transaction_validity`.
+        #[changed_in(2)]
         fn storage_keys_for_verifying_transaction_validity(
             account_id: opaque::AccountId,
             block_number: NumberFor<Block>,
             tx_era: Option<Era>,
         ) -> Result<Vec<Vec<u8>>, VerifyTxValidityError>;
 
+        /// Returns the storage keys of states accessed in the API `check_transaction_validity`
+        /// In v2, signer account id is optional to support getting storage keys for unsigned tx as well.
+        fn storage_keys_for_verifying_transaction_validity(
+            maybe_account_id: Option<opaque::AccountId>,
+            block_number: NumberFor<Block>,
+            tx_era: Option<Era>,
+        ) -> Result<Vec<Vec<u8>>, VerifyTxValidityError>;
+
+        /// New in v2.
+        /// Checks validity of bundle extrinsics. Internally calls `check_transaction_and_do_pre_dispatch`
+        fn check_bundle_extrinsics_validity(bundle_extrinsics: Vec<<Block as BlockT>::Extrinsic>, block_number: NumberFor<Block>,
```

Comment:
> This also make me think how big can the storage proof be

This depends on the exact storage accessed during the `pre_dispatch` of the illegal tx, but the size should be guarded by the excessive transaction check since `pre_dispatch` is also part of the execution of the tx.